File: Leetcode_go/46_Permutations.go
package main

func genPerm(ans [][]int, temp []int, visited []bool, nums []int) [][]int {
	if len(temp) == len(nums) {
		perm := make([]int, len(temp))
		copy(perm, temp)
		ans = append(ans, perm)
		return ans
	}
	for i := range nums {
		if visited[i] {
			continue
		}
		temp = append(temp, nums[i])
		visited[i] = true
		ans = genPerm(ans, temp, visited, nums)
		temp = temp[:len(temp)-1]
		visited[i] = false
	}
	return ans
}
func permute(nums []int) [][]int {
	visited := make([]bool, len(nums))
	return genPerm([][]int{}, []int{}, visited, nums)
}

File: Leetcode/Backtracking/46_Permutations.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
  void genPermutation(vector<vector<int>> &ans, vector<int> temp,
                      vector<int> &nums, vector<bool> &visited) {
    if (temp.size() == nums.size()) {
      ans.push_back(temp);
      return;
    }
    for (int i = 0; i < nums.size(); i++) {
      if (visited[i])
        continue;
      temp.push_back(nums[i]);
      visited[i] = true;
      genPermutation(ans, temp, nums, visited);
      temp.pop_back();
      visited[i] = false;
    }
  }
  vector<vector<int>> permute(vector<int> &nums) {
    vector<vector<int>> ans;
    vector<int> temp;
    vector<bool> visited(nums.size(), 0);
    genPermutation(ans, temp, nums, visited);
    return ans;
  }
};

int main() {
  vector<int> nums = {1, 2, 3};
  Solution s;
  auto ans = s.permute(nums);
  for (auto &it : ans) {
    for (auto &i : it)
      cout << i << " ";
    cout << endl;
  }
}
